Implement the concept of input generator specifications

sphuber opened the pull request:
The refactoring of the input generators for common workflows in this
commit comes from the following two needs:

 1. The exact signatures of the `get_builder` method of input generators
    needs to be customizable, as it has become clear that certain
    implementations need additional keyword arguments that may not be
    general enough to fit all implementations.

 2. The signature of the `get_builder` method of input generators needs
    to be introspectable. That is to say, it needs to be possible to
    programmatically determine which arguments it accepts, what the
    valid types for those arguments are, and additional information,
    such as help strings.

These requirements very closely match that of the `Process` base class
of `aiida-core` itself. It needs to allow subclasses to customize their
specification with respect to inputs and outputs. This functionality is
implemented through the concept of a process specification that build up
their input and output namespaces through `PortNamespace`'s and `Port`'s.

Instead of reinventing the wheel, we repurpose this concept and create a
base `InputGenerator` class. This base class defines the `define`
method, just like a `Process`, which gets an instance of the new class
`InputGeneratorSpec`. This is the analogue of the `ProcessSpec` and
allows to define the input specification of the input generator by
creating a, optionally nested, namespace of `InputGeneratorPort`s. These
define the keyword arguments that the `get_builder` method of the input
generator will accept.

By using the process specification functionality of `aiida-core`, it
becomes easy to define specification for input generators whose
validation is fully automated and whose specification is
programmatically introspectable. This will allow clients to
automatically build interfaces that interact with the input generators,
as well as automatically generate human-readable documentation.

sphuber commented:
@bosonie @giovannipizzi @mbercx Here the first rough implementation of the new input generator infrastructure which should make it easy to extend input generator signatures to add custom inputs (for example DFT-specific inputs) while still keeping everything automatically validated and introspectable. Some tests are failing but I am still working on those. There are a few minor technical things to tackle.

I also noticed that some plugins define support for certain arguments (for example support for electronic types) but then ignore their values. We should probably find an official way to define this if there is a difference between not supporting a type value or ignoring it all together.

bosonie commented:
I have several doubts about this approach. I'm sure this is due to the fact that I was not present at Friday's meeting. It might be worth reporting somewhere the outcomes of this meeting. Maybe here?
My major concern regards the fact that in this way we loose control on the common interface.  If you ask me in a "yes or no" question: "should the signatures of the `get_builder` method of input generators to be customizable?". My answer is "no". The `get_builder` signature should carry only the common inputs because these are the true common quantities needed to run a task (relaxation for instance). All the rest should be code details that the expert developer selects inside the protocol without the need for the general user to bother about them. The expert user can introspect the inputs of the `<Code>CommonRelaxWorkChain` and set everything he/she needs. Here it seems we are introducing a third level of inputs, in between the common inputs of the task and the code details. What is the use for it?  What is an example of an input that would fall in this category? The "smearing" was mentioned in previous meeting. This is something that is common to every DFT code, therefore it can become input of a future `DFTCommonRelaxationInputGenerator`, but it is going to be common to every implementation, not a customized input available for "some" codes.
Another doubt is the question on how to maintain this approach. It requires a lot more effort from developers to fix the interplay between the "protocol" string (mandatory) and this possible new inputs (they must be optional!). It's going to be a lot of "if" logic.
Sorry for the lengthy comment.

sphuber commented:
> Sorry for the lengthy comment.

Not at all, that is the whole point of me opening the PR to discuss.

> I have several doubts about this approach. I'm sure this is due to the fact that I was not present at Friday's meeting. It might be worth reporting somewhere the outcomes of this meeting. Maybe here?

The design here has nothing to do with the discussion of the last meeting. All that was discussed last meeting was to identify which potential additional parameters are _common_ to a subset of the current implementations, such as k-point distance, smearing width and type etc. We didn't discuss how this was going to be implemented.

> My major concern regards the fact that in this way we loose control on the common interface. If you ask me in a "yes or no" question: "should the signatures of the `get_builder` method of input generators to be customizable?". My answer is "no". The `get_builder` signature should carry only the common inputs because these are the true common quantities needed to run a task (relaxation for instance). All the rest should be code details that the expert developer selects inside the protocol without the need for the general user to bother about them. The expert user can introspect the inputs of the `<Code>CommonRelaxWorkChain` and set everything he/she needs. Here it seems we are introducing a third level of inputs, in between the common inputs of the task and the code details. What is the use for it? What is an example of an input that would fall in this category? The "smearing" was mentioned in previous meeting. This is something that is common to every DFT code, therefore it can become input of a future `DFTCommonRelaxationInputGenerator`, but it is going to be common to every implementation, not a customized input available for "some" codes.

The problem we are discussing consists really of two parts:

1. Do we want to allow for some _subvariants_ of certain common workflows that have a richer interface for certain implementations because they share a further set of common inputs, such as the DFT codes. 
2. If so, how do we implement this technically.

It seems to me that we had already answered the first question in the affirmative, and you also seem to not be against it. So the question is just about how to implement it. You propose that this should become a subclass that some implementations then use. But conceptually this is exactly what I am proposing. I can imagine this is maybe not immediately evident, since it is a PR in progress and so I haven't had the time to fully write out the documentation of what it would look like from the outside for a developer and user. So maybe we should meet in person soon so we can go over it.

> Another doubt is the question on how to maintain this approach. It requires a lot more effort from developers to fix the interplay between the "protocol" string (mandatory) and this possible new inputs (they must be optional!). It's going to be a lot of "if" logic.

That is exactly the point though. With the proposed design, the signature of the input generator is dynamic and so if an implementation doesn't support `smearing_type`, it doesn't define it in the spec and so it will never receive it and so it won't have to deal with it. On the other hand, the user will also automatically see that they shouldn't bother passing the `smearing_type` since it is not a supported input. This is exactly what you describe should be the case, unless I misunderstand what you meant

bosonie commented:
Thanks Sebastiaan. I could not go through the implementation properly...to have an explanation will be useful I believe. 
The misleading point for me was point 1. of your description. You say " it has become clear that certain implementations need additional keyword arguments that may not be general enough to fit all implementations.". I took this point as an indication that you wanted to allow each code-dependent implementation to set additional inputs in the signature of `get_builder`. I would oppose this. Instead, if you mean that this is just a technical detail that makes more clear the fact that some implementations do not support yet one of the inputs (for instance NWChen still do not support spin_type), then it is ok.
Do we have a way to make sure that code-dependent implementations do not set arbitrary inputs?
Still I have to see  the utility of your proposal compared to the current implementation, but I guess for it I need to dig deeper in the code.

sphuber commented:
> Do we have a way to make sure that code-dependent implementations do not set arbitrary inputs?

This is already possible. They can already add whatever input they want when they implement `get_builder`. Of course we can control what is added before accepting PRs. So again, the question for me is not how we prevent this from happening, because we currently aren't, but we also don't want to. The "common" interface already _is_ not fully common. There are already differences in what values are supported or what arguments are even supported. And we seemed to have been in agreement that we want to further extend this. I think this is fine, as long as:

* custom arguments are fully independent and do not complicate the implementation of other engines
* added arguments are automatically validated and introspectable, as all other common arguments

That is all that the implementation presented here changes. It just makes it easier to comply to this without a lot of additional custom code and copy-pasting.

sphuber commented:
To try and make the system I propose more easy to understand, I will describe here how it is actually used. Below, I will show how:

1. We can easily define the specification of common input generators.
2. How these common input generators can easily be subclassed and customized by implementations refining what arguments are supported, reducing overall code duplication and copy-pasting.
3. How the resulting input generators now have an interface that is programmatically introspectable, allowing interfaces to it, such as a CLI or a GUI,  to be automatically generated
4. How the programmatic specification allows human-readable documentation to be generated automatically.

### Defining a base input generator
As an example, let's take the `CommonRelaxInputGenerator`. With the new system, it defines in a concise way what the common arguments are:
```python
class CommonRelaxInputGenerator(InputGenerator, metaclass=abc.ABCMeta):
    """Input generator for the common relax workflow. """

    @classmethod
    def define(cls, spec):
        """Define the specification of the input generator.

        The ports defined on the specification are the inputs that will be accepted by the ``get_builder`` method.
        """
        super().define(spec)
        spec.input('structure', valid_type=plugins.DataFactory('structure'),
            help='The structure whose geometry should be optimized.'
        )
        spec.input('protocol', valid_type=ChoiceType(('fast', 'moderate', 'precise')), default='moderate')
        spec.input('spin_type', valid_type=SpinType, serializer=lambda x: SpinType(x), default=SpinType.NONE)
```
Note that I only show a few of the common arguments for brevity, but we would of course define all the common inputs here. Reusing the concept of process specification for the `input generator specification` not only gives us a powerful and well-tested system, but plugin developers are also already intimately familiar with it. This prevents them having to learn yet another interface and can simply reuse the knowledge they already have.

### Implementing the common input generator for a quantum engine
Since the common inputs for the generator have already defined by the base class, implementation for quantum engines have to merely update the specification with their specifics. For example, they typically only support a subset of the values of the `SpinType` enum that is accepted for the `spin_type` argument. The existing port can be updated without having to completely redefine it:
```python
class SiestaCommonRelaxInputGenerator(CommonRelaxInputGenerator):

    @classmethod
    def define(cls, spec):
        """Define the specification of the input generator."""
        super().define(spec)
        spec.inputs['spin_type'].valid_type = ChoiceType((SpinType.NONE, SpinType.COLLINEAR))
        spec.inputs['engines']['relax']['code'].valid_type = CodeType('siesta.siesta')
```
Note how the implementation takes minimal effort, which reduces the chance for mistakes, but also increases the homogeneity between all implementations. Furthermore, this system makes it easy to add new common arguments, which is a use case that we have identified we need for the second paper. For example, it is now easy to add a new argument that allows to change the smearing type:
```python
class SiestaCommonRelaxInputGenerator(CommonRelaxInputGenerator):

    @classmethod
    def define(cls, spec):
        """Define the specification of the input generator."""
        super().define(spec)
        spec.inputs['spin_type'].valid_type = ChoiceType((SpinType.NONE, SpinType.COLLINEAR))
        spec.inputs['engines']['relax']['code'].valid_type = CodeType('siesta.siesta')
        spec.input('smearing_type', valid_type=ChoiceType((SmearingType.GAUSSIAN, SmearingType.FERMI_DIRAC)))
```
Other implementations that also share support for this argument can simply add the same port to their specification. I want to stress that this method of adding new arguments _does not affect other implementations whatsoever_. If a user attempts to pass the `smearing_type` to the generator of an implementation that does not support it, an error will simply be thrown that this argument is not supported.

### Programmatically inspecting input generator specification
Now that we have covered how the system works for the developers, let's take a look at it from the user perspective. Thanks to the programmatic specification, it is possible to inspect it programmatically. Demonstration:
```python
In [15]: SiestaRelax = WorkflowFactory('common_workflows.relax.siesta')

In [16]: generator = SiestaRelax.get_input_generator()

In [17]: print(generator.spec().inputs)
{
    "electronic_type": {
        "default": "ElectronicType.METAL",
        "help": "The electronic character of the system.",
        "name": "electronic_type",
        "required": "False",
        "valid_type": "<enum 'ElectronicType'>"
    },
    "engines": {
        "_attrs": {
            "default": [],
            "dynamic": false,
            "help": "Inputs for the quantum engines",
            "required": "True",
            "valid_type": "None"
        },
        "relax": {
            "_attrs": {
                "default": [],
                "dynamic": false,
                "help": "Inputs for the quantum engine performing the geometry optimization.",
                "required": "True",
                "valid_type": "None"
            },
            "code": {
                "help": "The code instance to use for the geometry optimization.",
                "name": "code",
                "required": "True",
                "valid_type": "<class 'aiida.orm.nodes.data.code.Code'>"
            },
            "options": {
                "help": "Options for the geometry optimization calculation jobs.",
                "name": "options",
                "required": "False",
                "valid_type": "<class 'dict'>"
            }
        }
    },
    ...
    "threshold_stress": {
        "help": "A real positive number indicating the target threshold for the stress in eV/\u00c5^3. If not specified, the protocol specification will select an appropriate value.",
        "name": "threshold_stress",
        "required": "False",
        "valid_type": "<class 'float'>"
    }
}
```
I have again snipped out some of the ports for brevity. The `spec` method of the inputs generator returns the input specification which contains all the inputs that have been defined which specify their valid type, help string, etc. This helps the discoverability not just for humans, but also for programs that can automatically write interfaces for it.

### Automatically generated documentation
This is one of the more powerful features of the proposed system. I have already coded a Sphinx extension that takes an input generator implementation and automatically converts its specification in a human-readable HTML document. All that one has to write in terms of documentation is the following directive:
```rst
SIESTA
------

.. common-input-generator:: SiestaCommonRelaxInputGenerator
    :module: aiida_common_workflows.workflows.relax.siesta.generator
```
With the extension that I wrote, this will automatically be converted into:
![auto-docs](https://user-images.githubusercontent.com/6992332/135164897-a7acb2e3-ce10-4cda-8131-d6401bd1ebfb.png)

This hopefully demonstrates how the proposed system would provide high flexibility and easy reuse for developers, while providing high usability for users through introspectability and automated documentation generation.

giovannipizzi commented:
Hi @sphuber thanks! I had originally similar concerns as @bosonie but after looking into your more detailed explanation, I like the idea so I'm OK with approving it!

PS: To clarify: the concern was about letting people change at will the possible inputs of the input generator. However, without the changes of this PR, this would still be possible. Theferore, this PR doesn't change much on that aspect (but improves on many others).
Maybe three comments:
- in the docs, clarify explicitly in a box that input generator implementations should not add more inputs than those defined in the common specifications
- If they want to allow for more customisations, I'm leaning more and more for the idea of providing "lambda override functions" that are code specific and can take the output of the common input generator and modify it as needed. An implementation can provide as many as they want, with appropriate documentation, and these would work much better when the common relaxation workflow is used within some other common workflow (e.g. EOS): with custom inputs, the EOS would not work; but with a (list of) lambda overrides, one can just pass them to the EOS as overrides and these can be applied to each step. [this is probably for a more general discussion, I will add some comments in #206 for now]
- since, with this, inputs are machine-readable, I suggest you add an issue, when you merge this, about automatic validation (e.g. in GitHub actions) that all implementations provided here satisfy some requirements (do not define new inputs, have valid values, ...)

espenfl commented:
@sphuber Thanks for taking the time for the detailed explanation. I think this makes a lot of sense from all perspectives. Particularly code maintenance, extendibility and documentation. The latter I find particularly important as this tends to be put all the way down on the todo list for some developers. Having this functionality would give at least some basic documentation to users. I support this PR.

sphuber commented:
> in the docs, clarify explicitly in a box that input generator implementations should not add more inputs than those defined in the common specifications

So this would be docs for developers that would want to add an implementation, correct? I agree that we would probably need to write documentation for developers at some point as the code is becoming more extensive. Still, since we anyways go through PRs, we can always make sure before merging that their specs respect the common specification.

That being said, I think we need to define a bit more concretely what we mean with the "common interface". Both @bosonie and yourself indicated that the prospect of implementations not respecting this was the initial source of your hesitation for these changes. However, as I already mentioned before, at the same time we all agree that we probably _need_ to allow to customize the common interface. Or at the very least extend it, as we are discussing for the DFT based codes allowing to add additional arguments to specify the desired k-points distance, smearing etc. So if this changing of the interface _is_ allowed, then what kind of changes should not be allowed?

On top of that, as you also remarked already, there is currently nothing that stops an implementation from not respecting the common interface anyway. In fact, there already _are_ implementations that don't respect it (not a fault of the developers, but simply because their codes cannot support a part of the common interface). Example is the implementation for Gaussian and Orca that completely ignore the `electronic_type`. Currently they still define the `METAL` and `INSULATOR` values as "supported", as in they accept them to the builder, but then the values are simply ignored. This is not really a good approach and I actually propose to change it by dropping the `electronic_type` from the spec of their input generators. This way a user won't accidentally pass the value which will be silently swallowed, but instead will get an error. Through introspection it will also be perfectly clear that this argument is not supported.

Of course, this makes that their interface "differs" from that of the others, but I think this is ok. We have already established that we need flexibility. We should remember what the goal of the common interface is: it is a tool to make using different quantum engines as easy as possible for common tasks. It is _not_, or I think it _should_ not be a tool that has a perfectly identical interface for all implementations. This is simply not realistic nor really the goal of the project.

> If they want to allow for more customisations, I'm leaning more and more for the idea of providing "lambda override functions" that are code specific and can take the output of the common input generator and modify it as needed.

I am a bit confused. I thought the discussion of allowing lambdas for overrides related to your discussion in #206 which concerns letting users override the values of certain builder inputs. _Not_ for developers to override certain arguments of the `get_builder` interface?

Anyway, as you said, this is probably a different discussion and we should continue it in #206. I will give my thoughts there later.

> since, with this, inputs are machine-readable, I suggest you add an issue, when you merge this, about automatic validation (e.g. in GitHub actions) that all implementations provided here satisfy some requirements (do not define new inputs, have valid values, ...)

This is a good point and is something that I was going to address. Note that this is already in place in a sense, however, it is manually copy-pasted for each implementation. I already added for each implementation the test that for the `electronic_type`, `spin_type` and `relax_type` values they claim they support, that the `get_builder` returns a builder successfully. This can later just be improved by having these tests written "dynamically" without copy-pasting them literally which is a pain for maintenance.

>  This new approach would be only for the common workflow interfaces, or also for the common workflows like the EOS?
>  if also for the EOS, will it be possible to pass a lambda function, in case you accept my proposal in  #206?

I am not sure what you mean. This PR concerns the _input generators_ of the common _base_ workflows, such as the common relax workflow. For those  _base_ workflows we need an input generator, for which I introduced here a more rigorous method of defining its input specification. The _composite_ common workflows, such as the EOS and DC workflows, only have an input specification, which is the usual process specification since they are simply `WorkChain`. They don't need an input generator and so they are also are not concerned by this PR.

sphuber commented:
I have refactored the PR in two commits and added a third with some generic tests on consistency of the implementation of the input generator specifications. This is now ready for final approval and merging.

giovannipizzi commented:
> So this would be docs for developers that would want to add an implementation, correct? 

Yes

> we probably need to allow to customize the common interface. Or at the very least extend it, as we are discussing for the DFT based codes allowing to add additional arguments to specify the desired k-points distance, smearing etc.

> I am a bit confused. I thought the discussion of allowing lambdas for overrides related to your discussion in #206 which concerns letting users override the values of certain builder inputs. Not for developers to override certain arguments of the get_builder interface?

> Currently they still define the METAL and INSULATOR values as "supported", as in they accept them to the builder, but then the values are simply ignored. This is not really a good approach and I actually propose to change it by dropping the electronic_type from the spec of their input generators. 

I try reply to all three points here. First, I think that if we define the concept of sub-classes, the additional arguments (k-points distance, smearing, ...) could be defined for all codes that decide to implement that specific subclass. In a similar way, we could decide an even "super" class for the relaxation that has only very minimal inputs (no electronic_type, and maybe one could also have a very top-level one that just has the mandatory inputs and the thresholds, and nothing else, for e.g. very simple Lennard-Jones potentials). Finally, my point with the overrides (whose discussion, indeed, belongs to #206) is: *if* we decide that having lambda overrides is the way to go, *and* if we turn out implementing libraries of lambdas for each implementation, then we can decide to be strict on what extensions can be implemented to the input generators (read: almost none w.r.t. the base class) and any customisation should go via lambdas.

I'd need to think if, from a code maintainability point of view, this is easy or becomes easily too complex. But *if* we need to define lambdas anyway to support custom use cases, at that point it might be worth re-using the same concept also for any customisation.

Moreover, probably if an implementation adds custom inputs, these might not be reachable from the EOS-like workflows? (I note however how maybe this will not be the case after this PR, as you can expose the whole set of ports).

Anyway, this is up for discussion and maybe we realise your suggested approach (allow, instead, to add custom inputs) is better (and I agree that after merging this, it's a good moment to revisit all implementations and check if there are inconsistencies and fix them).

> The composite common workflows, such as the EOS and DC workflows, only have an input specification, which is the usual process specification since they are simply WorkChain. They don't need an input generator and so they are also are not concerned by this PR.

Right, thanks for the clarification! Just for a final double check (for #206): in the specs, it would be allowed to define non_db inputs that are lambdas?

giovannipizzi commented:
@sphuber I just hit 'update branch' but maybe you wanted to keep separate commits and not squash? Sorry if that was the plan (let me know if you want me to do anything or you want to rebase).

Maybe we should go ahead and merge this (so we can move to the definition of inputs for the smearing etc.)? @bosonie do you think this is OK or do you still want to discuss this?

sphuber commented:
@giovannipizzi good idea to merge, but indeed this was intended not to be squashed. I have reset the merge so we are all good and will merge then
